### PR TITLE
SM: smoother co-op, basepatch internal improvements

### PR DESCRIPTION
**Pull request by strotlog**
- Remove hard-coded ROM addresses from world generation. If an AP SM data table's ROM location gets moved at all within the basepatch in a future revision, the AP generator will continue writing to that table at its new location without needing any change to world code.
    * Easier to make small basepatch changes, or simultaneous-ish changes from multiple contributors.
    * Also makes the world code easier to read and see how it works.
    * Note: **SNIClient** should continue to read ROM/SRAM/WRAM addresses using hard-coded addresses for now, since there's no system for SNIClient to fetch the symbol addresses used by the AP server's SM world revision. Nothing is currently being moved that SNIClient touches.
- Allow full 200 player names (SM can never need to reference more) + "Archipelago" fallback player name

Incorporates my proposed refactor and improvements at https://github.com/lordlou/SMBasepatch/pull/1 into AP. I think we're only waiting on merging that PR there in case there's feedback for it that comes in here on this downstream PR.

Direct benefits from that PR:

- Co-op items (and all other "remote items" use cases) now collect items on touch instead of after network delay, a more natural/intuitive experience for co-op gameplay
- 0.3.3 fix for broken remote items (https://github.com/ArchipelagoMW/Archipelago/pull/592) backported to pre-0.3.3 clients
- Reduces hidden dependencies on VARIA randomizer code
- Reduces Bank 0x84 freespace usage for multiworld from 34% of freespace down to 13%, allowing room for the latest features over at VARIA randomizer (totaling 76% of the bank's freespace) to be integrated in the future
- Possible from a ROM perspective to add AP support for popular romhacks of SM (though no AP logic has been developed for them)

This will require review by @lordlou who has already been quite helpful! Sorry it took me a while lordlou. As far as testing, multiworlds have been completed at various stages of development with all known issues fixed, and I think a little more coverage would be great. DM me if you need a buddy pre-RC.

(Previously said this fixed an issue with starting items, but it was my own issue! Sorry about that)

### CODE REVIEW

**Review by lordlou — COMMENTED** [review]
I did a quick test and I found that remote items graphic is missing. I can collect the item fine but its like a "Nothing" item visually.

Apart from that, its looking great!

File: worlds/sm/__init__.py [hunk 1]
```diff
@@ -269,158 +266,232 @@ def convertToROMItemName(self, itemName):
...
+            Utils.local_path(       "worlds", "sm", "data", "SMBasepatch_prebuilt", "multiworld-basepatch.ips"))
+        romPatcher.applyIPSPatch(
+            Utils.local_path("lib", "worlds", "sm", "data", "SMBasepatch_prebuilt", "variapatches.ips") if
+            Utils.is_frozen() else
+            Utils.local_path(       "worlds", "sm", "data", "SMBasepatch_prebuilt", "variapatches.ips"))
+
+        symbols = get_sm_symbols(
+            Utils.local_path("lib", "worlds", "sm", "data", "SMBasepatch_prebuilt", "sm-basepatch-symbols.json") if
+            Utils.is_frozen() else
+            Utils.local_path(       "worlds", "sm", "data", "SMBasepatch_prebuilt", "sm-basepatch-symbols.json"))
```
[the remark this answers is not in the document]
Comment (strotlog) [reply worlds/sm/__init__.py:290]: Great idea, will add!

File: worlds/sm/__init__.py [hunk 2]
```diff
@@ -269,158 +266,232 @@ def convertToROMItemName(self, itemName):
...
 
+        openTourianGreyDoors = {0x07C823 + 5: [0x0C], 0x07C831 + 5: [0x0C]}
+        romPatcher.applyIPSPatchDict({'openTourianGreyDoors': openTourianGreyDoors})
+
+
         # set rom name
         # 21 bytes
         from Main import __version__
-        self.romName = bytearray(f'SM{__version__.replace(".", "")[0:3]}_{self.player}_{self.world.seed:11}\0', 'utf8')[:21]
+        self.romName = bytearray(f'SM{__version__.replace(".", "")[0:3]}_{self.player}_{self.world.seed:11}', 'utf8')[:21]
```
[the remark this answers is not in the document]
Comment (strotlog) [reply worlds/sm/__init__.py:404]: I think I made this change a really long time ago on my local copy, it's not a big deal if you want it removed. I don't think null termination does any harm or good either way.

Maybe what I had noticed was that if the player ID is 3 digits long, then the null terminator will get truncated anyway, as the string will be this long leading up to it (21 chars):
SM034_999_00112233441
Or to grab someone's actual seed from the current 600 world async:
SM032_102_24586044026 (no null terminator)

Everything that reads the SNES title should be aware that its max length is 21 btyes (some games run right up to the last byte with text, some pad the right side with spaces (0x20) but include no \0 null terminator).

[the remark this answers is not in the document]
Comment (strotlog) [reply worlds/sm/__init__.py:400]: I forgot to look at the next line. In the case that the ROM title is less than 21 characters, the next line adds null terminators, so I considered the literal '\0' to be a no-op. (As well as misleading for someone reading the code and deciding to expect a null terminator, in the 21-character case)

[the remark this answers is not in the document]
Comment (lordlou) [reply worlds/sm/__init__.py:400]: Indeed, good then.

**Comment by lordlou** [comment]
I reviewed this PR and I did a quick test and everything seems good.

**Review by Berserker66 — APPROVED** [review]

**Pull request merged**